### PR TITLE
Implement client labels

### PR DESCRIPTION
**Pull request by ihsandemir**
Implemented client labels.

### CODE REVIEW

File: hazelcast-client/src/main/java/com/hazelcast/client/config/ClientConfig.java [hunk 1]
```diff
@@ -1014,39 +1015,20 @@ public QueryCacheConfig getOrNullQueryCacheConfig(String mapName, String cacheNa
...
      *
-     * @param key   the key of the attribute
-     * @param value the value of the attribute
-     * @return configured {@link com.hazelcast.client.config.ClientConfig} for chaining
+     * @param label The label to be added.
      */
-    public ClientConfig setAttribute(String key, String value) {
-        attributes.put(key, value);
-        return this;
+    public void addLabel(String label) {
```
Comment (sancar) [inline hazelcast-client/src/main/java/com/hazelcast/client/config/ClientConfig.java]: Can you put ClientConfig as a return value? Also please mention  {@link com.hazelcast.core.Client} on the javadoc.

File: hazelcast-spring/src/main/resources/hazelcast-spring-3.12.xsd [hunk 2]
```diff
@@ -1925,6 +1925,7 @@
                         <xs:element name="instance-name" type="xs:string" minOccurs="0" maxOccurs="1"/>
                         <xs:element name="properties" type="properties" minOccurs="0" maxOccurs="1"/>
                         <xs:element name="attributes" type="attributes" minOccurs="0" maxOccurs="1"/>
```
Comment (sancar) [inline hazelcast-spring/src/main/resources/hazelcast-spring-3.12.xsd]: Attributes needs to be removed

File: hazelcast/src/test/java/com/hazelcast/client/protocol/compatibility/BinaryCompatibilityFileGenerator.java [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
```
Comment (sancar) [inline hazelcast/src/test/java/com/hazelcast/client/protocol/compatibility/BinaryCompatibilityFileGenerator.java]: It seems that we need to update header on protcol

File: hazelcast/src/test/java/com/hazelcast/client/protocol/compatibility/BinaryCompatibilityFileGenerator.java [hunk 4]
```diff
@@ -16,40 +16,57 @@
...
-import java.util.HashMap;
-import java.util.LinkedList;
-import java.util.List;
-import java.util.ListIterator;
-import java.util.Map;
-
-import static org.junit.Assert.assertTrue;
-import static com.hazelcast.client.protocol.compatibility.ReferenceObjects.*;
+
+import static com.hazelcast.client.protocol.compatibility.ReferenceObjects.aBoolean;
```
Comment (sancar) [inline hazelcast/src/test/java/com/hazelcast/client/protocol/compatibility/BinaryCompatibilityFileGenerator.java]: Can you make sure that your IDE does not arrange the imports?

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AuthenticationCustomCredentialsMessageTask.java [hunk 5]
```diff
@@ -68,12 +68,12 @@ protected String getClientType() {
             clientName = parameters.clientName;
         }
 
-        if (parameters.attributesExist) {
-            Map<String, String> map = new HashMap<String, String>();
-            for (Map.Entry<String, String> attribute : parameters.attributes) {
-                map.put(attribute.getKey(), attribute.getValue());
+        if (parameters.labelsExist) {
```
Comment (sancar) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AuthenticationCustomCredentialsMessageTask.java]: We can replace the iteration with 
labels = Collections.unmodifiableSet(new HashSet<String>(parameters.labels));

File: hazelcast/src/main/java/com/hazelcast/internal/management/dto/ClientEndPointDTO.java [hunk 6]
```diff
@@ -56,11 +56,11 @@ public JsonObject toJson() {
         root.add("address", address);
         root.add("clientType", clientType);
         root.add("name", name);
-        JsonObject attrObject = Json.object();
-        for (Map.Entry<String, String> entry : attributes.entrySet()) {
-            attrObject.add(entry.getKey(), entry.getValue());
+        JsonObject labelsObject = Json.object();
+        for (String label : labels) {
+            labelsObject.add("label", label);
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/internal/management/dto/ClientEndPointDTO.java]: This serialization is not correct. It leads to duplicate keys in JSON:
```
{
  "labels": {
    "label": "firstLabel-hzclient-1",
    "label": "secondLabel-hzclient-1"
  }
}
```

An array of strings should be used, so the result would be:
```
{
  "labels": [
    "firstLabel-hzclient-1",
    "secondLabel-hzclient-1"
  ]
}
```

File: hazelcast/src/main/java/com/hazelcast/internal/management/dto/ClientEndPointDTO.java [hunk 7]
```diff
@@ -70,11 +70,11 @@ public void fromJson(JsonObject json) {
...
         clientType = getString(json, "clientType");
         name = getString(json, "name");
-        JsonObject attrObject = JsonUtil.getObject(json, "attributes");
-        attributes = new HashMap<String, String>();
-        for (JsonObject.Member member : attrObject) {
-            String value = member.getValue().asString();
-            attributes.put(member.getName(), value);
+        JsonObject labelsObject = JsonUtil.getObject(json, "labels");
+        labels = new HashSet<String>();
+        for (JsonObject.Member labelMember : labelsObject) {
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/internal/management/dto/ClientEndPointDTO.java]: Most likely, deserialization should be updated too (see comment about duplicate keys in JSON).

File: hazelcast/src/main/java/com/hazelcast/internal/management/dto/ClientEndPointDTO.java (hunk 6 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/management/dto/ClientEndPointDTO.java]: Good catch

Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/internal/management/dto/ClientEndPointDTO.java]: Changed

File: hazelcast/src/main/java/com/hazelcast/internal/management/dto/ClientEndPointDTO.java (hunk 7 above)
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/internal/management/dto/ClientEndPointDTO.java]: Corrected.

**Comment by ihsandemir** [comment]
run-lab-run

**Comment by sancar** [comment]
run-lab-run

**Comment by sancar** [comment]
run-lab-run

**Pull request merged**